### PR TITLE
Add check if a certain key/value cached.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Add functionality to check if key in the cache.
Add tests.
    
Resolves: OLPEDGE-2110
    
Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/src/cache/DiskCache.cpp [hunk 1]
```diff
@@ -317,6 +317,18 @@ bool DiskCache::Get(const std::string& key,
...
 }
 
+bool DiskCache::Exist(const std::string& key) {
+  if (!database_) {
+    OLP_SDK_LOG_ERROR(kLogTag, "Get: Database is not initialized");
+    return false;
+  }
+  leveldb::ReadOptions options;
+  options.verify_checksums = check_crc_;
+  auto iterator = NewIterator(options);
```
Comment (kzvieriev) [inline olp-cpp-sdk-core/src/cache/DiskCache.cpp:327]: Set `fill_cache` option to false, so we don't load data.

File: olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp [hunk 2]
```diff
@@ -303,6 +303,21 @@ bool DefaultCacheImpl::RemoveKeysWithPrefix(const std::string& key) {
...
+    auto iterator = mutable_cache_lru_->FindNoPromote(key);
+    return (iterator != mutable_cache_lru_->end());
+  }
+  if (mutable_cache_ && mutable_cache_->Exist(key)) {
+    return true;
+  }
+  if (protected_cache_ && protected_cache_->Exist(key)) {
+    return true;
+  }
+  return false;
```
Comment (kzvieriev) [inline olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp:318]: I believe, need to check memory cache too.
Also no need to check mutable cache if key not found in the lru.

File: olp-cpp-sdk-core/include/olp/core/cache/KeyValueCache.h [hunk 3]
```diff
@@ -121,6 +121,14 @@ class CORE_API KeyValueCache {
...
    */
   virtual bool RemoveKeysWithPrefix(const std::string& prefix) = 0;
+  /**
+   * @brief Check if key is in the cache.
+   *
+   * @param key The key for the value.
+   *
+   * @return True if the key/value cached ; false otherwise.
+   */
+  virtual bool Exist(const std::string& key) = 0;
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/cache/KeyValueCache.h:131]: Lets call it Contains().
NOTE: This is a breaking change! This means all user code will not compile as there is a new abstract API that is not overridden. You need to provide a default impl to not break API.

```suggestion
  virtual bool Contains(const std::string& key) {
      CORE_UNUSED(key);
   };
```

File: olp-cpp-sdk-core/include/olp/core/cache/KeyValueCache.h [hunk 4]
```diff
@@ -121,6 +121,14 @@ class CORE_API KeyValueCache {
    * @return True if the values are removed; false otherwise.
    */
   virtual bool RemoveKeysWithPrefix(const std::string& prefix) = 0;
+  /**
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/cache/KeyValueCache.h:124]: Seems to be a recurring issue.
Please separate methods with an empty newline.

File: olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp [hunk 5]
```diff
@@ -303,6 +303,30 @@ bool DefaultCacheImpl::RemoveKeysWithPrefix(const std::string& key) {
   return true;
 }
 
+bool DefaultCacheImpl::Exist(const std::string& key) {
```
Comment (andescu) [inline olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp:306]: Please separate individual blocks by a newline.

File: olp-cpp-sdk-core/src/cache/DiskCache.h [hunk 6]
```diff
@@ -141,6 +141,8 @@ class DiskCache {
   bool RemoveKeysWithPrefix(const std::string& prefix,
                             uint64_t& removed_data_size);
 
+  bool Exist(const std::string& key);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/cache/DiskCache.h:144]:
```suggestion
  bool Contains(const std::string& key);
```

File: olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp [hunk 7]
```diff
@@ -303,6 +303,30 @@ bool DefaultCacheImpl::RemoveKeysWithPrefix(const std::string& key) {
...
+  std::lock_guard<std::mutex> lock(cache_lock_);
+  if (!is_open_) {
+    return false;
+  }
+  if (memory_cache_ && !(memory_cache_->Get(key).empty())) {
+    return true;
+  }
```
Comment (kzvieriev) [inline olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp:313]: Can we add Exist() method for memory cache too, so we don't copy data from it?

File: olp-cpp-sdk-core/src/cache/InMemoryCache.cpp [hunk 8]
```diff
@@ -105,6 +105,22 @@ void InMemoryCache::RemoveKeysWithPrefix(const std::string& key_prefix) {
   }
 }
 
+bool InMemoryCache::Contains(const std::string& key) {
+  std::lock_guard<std::mutex> lock{mutex_};
+  auto it = item_tuples_.Find(key);
+  if (it != item_tuples_.end()) {
+    auto expiry_time = std::get<1>(it.value());
+    if (expiry_time < time_provider_()) {
+      PurgeExpired(expiry_time);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/cache/InMemoryCache.cpp:114]: Why do we erase during checking if it exists? We should only delete if we add something new.

**Pull request merged**